### ISSUE TITLE
ERR_HTTP2_GOAWAY_SESSION error

### ISSUE THREAD

**Issue by rauchg** [grantila/fetch-h2#19]
We somewhat frequently see this error in `now-cli`, which is currently using `fetch-h2`:
https://github.com/zeit/now-cli/blob/canary/package.json#L138

```
> Error! An unexpected error occurred!
  Error [ERR_HTTP2_GOAWAY_SESSION]: New streams cannot be created after receiving a GOAWAY
    at ClientHttp2Session.request (internal/http2/core.js:1364:13)
    at O.Promise.race.e.get.then (/snapshot/repo/dist/now.js:192:7125) Error [ERR_HTTP2_GOAWAY_SESSION]: New streams cannot be created after receiving a GOAWAY
    at ClientHttp2Session.request (internal/http2/core.js:1364:13)
    at O.Promise.race.e.get.then (/snapshot/repo/dist/now.js:192:7125)
```

Any idea how we could prevent / re-try this?

**Comment by grantila** [comment]
Oh no, this day came. And it was inevitable. Guess I'll have to implement session failover for streams following a GOAWAY.

I'm not sure why your server sends GOAWAY "somewhat frequently", I realize this could be caused by the server wanting to close the connection for some downstream maintenance. But often sounds strange:

[Spec:](https://http2.github.io/http2-spec/#GOAWAY)
> The GOAWAY frame (type=0x7) is used to initiate shutdown of a connection or to signal serious error conditions. GOAWAY allows an endpoint to gracefully stop accepting new streams while still finishing processing of previously established streams. This enables administrative actions, like server maintenance.

Anyway, I suppose this should be handled gracefully by simply opening up a new session and using that for further streams. That does mean quite some restructuring and a regression tests, but I'm on it. This is one of the things that should "just work" in a library like this.

**Comment by grantila** [comment]
@rauchg do you get this on non-request-body methods (HEAD, GET) or any of the other (PUT, POST, ...)?

I'm implementing automatic retries (on race-conditioned GOAWAY's) and session fail-over on mid-request GOAWAY's. It's the "same thing" for an end-user, but different for a client implementation. In fact, it's pretty nasty:

Since a race-conditioned GOAWAY (one that is sent from the server at the same time as a request body has been partially sent by the client) causes the client to not have the complete body when it is signalled that the stream is closed and should be retried on a new session, the client can't properly retry.
Hence, I'll only implement support for this edge case race-condition GOAWAY for HEAD/GET for now.

H2 is more complex than one can imagine...

**Comment by grantila** [comment]
(after all, an edge case is an edge case and shouldn't be common at all)

**Comment by grantila** [comment]
(never mind the 0.2.7 mishap)
This is fixed in 1.0.2, just patch-bump and you should be fine or let me know.
